### PR TITLE
library: add PreferencesWindow entry

### LINKED ISSUE #288: AdwPreferencesWindow
https://gnome.pages.gitlab.gnome.org/libadwaita/doc/main/class.PreferencesWindow.html

Should not be too complicated, it's really about the window itself and include

* A couple of simple rows 
* A couple of groups
* Subpages (see libadwaita demo)

BTW the Workbench Library entry is using (abusing) `AdwPreferencesWindow`

### PR DESCRIPTION
**Pull request by halfmexican**
I'm still not sure why I am getting this error.
image

closes #288

### CODE REVIEW

File: src/Library/demos/Preferences Window/main.blp [hunk 1]
```diff
@@ -0,0 +1,160 @@
+using Gtk 4.0;
+using Adw 1;
+
+Adw.PreferencesWindow pref_window {
+  default-width: 800;
+  default-height: 600;
+  title: "My App Preferences";
```
Comment (sonnyp) [inline src/Library/demos/Preferences Window/main.blp:7]:
```suggestion
  title: _("Preferences");
```

please always use `_` for display strings

File: src/Library/demos/Preferences Window/main.blp [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+          halign: center;
+          valign: center;
+        }
+      }
+    }
+
+    Adw.PreferencesGroup {
+      title: "Text Settings";
+      description: "Customize the appearance of text in the application.";
+      //
```
Comment (sonnyp) [inline src/Library/demos/Preferences Window/main.blp:33]: ?

File: src/Library/demos/Preferences Window/main.blp [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+
+      Adw.ActionRow {
+        title: "Font Size";
+
+        [suffix]
+        Gtk.SpinButton{
```
Comment (sonnyp) [inline src/Library/demos/Preferences Window/main.blp:39]:
```suggestion
        Gtk.SpinButton {
```

File: src/Library/demos/Preferences Window/main.blp [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+          halign: center;
+          valign: center;
+          adjustment:
+          Gtk.Adjustment size_adj {
```
Comment (sonnyp) [inline src/Library/demos/Preferences Window/main.blp:43]:
```suggestion
          adjustment: Gtk.Adjustment {
```

this object id is not used

File: src/Library/demos/Preferences Window/main.blp [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+          halign: center;
+          valign: center;
+          dialog: ColorDialog {};
+        }
+      }
+
+      LinkButton {
+        label: "API Reference";
+        uri: "https://gnome.pages.gitlab.gnome.org/libadwaita/doc/1.3/class.PreferencesWindow.html";
+      }
```
Comment (sonnyp) [inline src/Library/demos/Preferences Window/main.blp:66]: please add a margin above, similar to other demos

I don't think we'll have separate entries for related widgets

We can add links for

* PreferencesWindow API Reference
* PreferencesPage API Reference
* PreferencesGroup API Reference
* PreferencesRow API Reference

File: src/Library/demos/Preferences Window/main.blp [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+Adw.PreferencesWindow pref_window {
+  default-width: 800;
+  default-height: 600;
+  title: "My App Preferences";
+
+  Adw.PreferencesPage appearance_page {
+    title:"Appearance";
+    icon-name:"brush-monitor-symbolic";
+
+    Adw.PreferencesGroup{
```
Comment (sonnyp) [inline src/Library/demos/Preferences Window/main.blp:13]:
```suggestion
    title: "Appearance";
    icon-name: "brush-monitor-symbolic";

    Adw.PreferencesGroup {
```

please watch out for formatting

File: src/Library/demos/Preferences Window/main.blp [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+        title: "Font Size";
+
+        [suffix]
+        Gtk.SpinButton{
+          halign: center;
+          valign: center;
+          adjustment:
+          Gtk.Adjustment size_adj {
+            lower: 5;
+            upper: 999;
```
Comment (sonnyp) [inline src/Library/demos/Preferences Window/main.blp:45]: this is not a great examplem maybe 20 or whatever?

File: src/Library/demos/Preferences Window/main.blp [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+      }
+
+      Adw.ActionRow {
+        title: "Font Color";
+
+        [suffix]
+        Gtk.ColorDialogButton{
+          halign: center;
+          valign: center;
+          dialog: ColorDialog {};
```
Comment (sonnyp) [inline src/Library/demos/Preferences Window/main.blp:59]: is dialog needed? I would expect this to work out of the box

File: src/Library/demos/Preferences Window/main.blp [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+  Adw.PreferencesPage behavior_page {
+    title:_("Behavior");
+    icon-name:"settings-symbolic";
+
+
+    Adw.PreferencesGroup{
+      title:_("Interaction Settings");
+      description:"Change how the app behaves during user interaction.";
+
+      //Adw.PreferencesGroup can have suffix widgets like Adw.ActionRows
```
Comment (sonnyp) [inline src/Library/demos/Preferences Window/main.blp:79]:
```suggestion
      // Adw.PreferencesGroup can have suffix widgets like Adw.ActionRows
```

File: src/Library/demos/Preferences Window/main.js [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+dm_switch.connect("state-set", (widget, state) => {
+  // When the Switch is toggled, set the color scheme
+  if (state) {
+    style_manager.color_scheme = Adw.ColorScheme.FORCE_DARK;
+  } else {
+    style_manager.color_scheme = Adw.ColorScheme.FORCE_LIGHT;
+  }
+});
+
+//Preferences windows can display subpages
```
Comment (sonnyp) [inline src/Library/demos/Preferences Window/main.js:23]: wdyt of

```suggestion
// PreferencesWindow can display subpages
```

File: src/Library/demos/Preferences Window/main.js [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+const dm_switch = workbench.builder.get_object("dm_switch");
+const subpage = workbench.builder.get_object("subpage");
+const subpage_row = workbench.builder.get_object("subpage_row");
+const subpage_button = workbench.builder.get_object("subpage_button");
+const toast_button = workbench.builder.get_object("toast_button");
+const style_manager = Adw.StyleManager.get_default();
+
+dm_switch.active = style_manager.dark;
+
+dm_switch.connect("state-set", (widget, state) => {
```
Comment (sonnyp) [inline src/Library/demos/Preferences Window/main.js:14]: this is not the right thing to connect to

please use `notify:active` signal and `switch.active` property

File: src/Library/demos/Preferences Window/main.blp [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+        Gtk.Image{
+          icon-name: "arrow1-right-symbolic";
+        }
+      }
+    }
+  }
+}
+
+Adw.StatusPage subpage {
+  title: "Preferences windows can have subpages like this one";
```
Comment (andyholmes) [inline src/Library/demos/Preferences Window/main.blp:152]: I'd suggest changing this to "Custom Subpage" or something like that, and move the existing label to the `description` property.

**Review by sonnyp — CHANGES_REQUESTED** [review]
Looks great.

**Comment by sonnyp** [comment]
BTW I don't get these errors. Can you see if you can reproduce after `flatpak update` and restarting Workbench?

File: src/Library/demos/Preferences Window/main.blp (hunk 3 above)
Comment (sonnyp) [reply src/Library/demos/Preferences Window/main.blp:39]: please fix the formatting to be consistent

File: src/Library/demos/Preferences Window/main.blp (hunk 8 above)
Comment (halfmexican) [reply src/Library/demos/Preferences Window/main.blp:59]: image
If I don't include a dialog i get this error

**Comment by halfmexican** [comment]
I still get the errors even after flatpak update

**Review by sonnyp — APPROVED** [review]
LGTM

I fixed a couple of minor issues https://github.com/sonnyp/Workbench/pull/297/commits/d0ada53672258ce416cd7beba446e2dd83e10aff

**Comment by sonnyp** [comment]
@andyholmes do you want to have an other look?

**Comment by andyholmes** [comment]
> @andyholmes do you want to have an other look?

Looks good to me! Nice and clean, with lots of examples.

**Review by andyholmes — APPROVED** [review]

**Pull request merged**